pkg/client/cache/reflector.go

```diff
@@ -160,16 +164,50 @@ var (
 // required, and a cleanup function.
 func (r *Reflector) resyncChan() (<-chan time.Time, func() bool) {
 	if r.resyncPeriod == 0 {
+		r.nextResync = time.Time{}
 		return neverExitWatch, func() bool { return false }
 	}
 	// The cleanup function is required: imagine the scenario where watches
 	// always fail so we end up listing frequently. Then, if we don't
 	// manually stop the timer, we could end up with many timers active
 	// concurrently.
+	r.nextResync = time.Now().Add(r.resyncPeriod)
 	t := time.NewTimer(r.resyncPeriod)
 	return t.C, t.Stop
 }
 
+// We want to avoid situations when periodic resyncing is breaking the TPC
+// connection.
+// If response`s body is not read to completion before calling body.Close(),
+// that TPC connection will not be reused in the future - see #15664 issue
+// for more details.
+// Thus, we set timeout for watch requests to be smaller than the remaining
+// time until next periodic resync and force resyncing ourself to avoid
+// breaking TCP connection.
+func (r *Reflector) timeoutForWatch() *int64 {
+	randTimeout := time.Duration(float64(5*time.Minute) * (rand.Float64() + 1.0))
+	// Set timeout so that watch request finishes no later than 'threshold'
+	// before next planned resync.
+	threshold := time.Second
+	timeout := r.nextResync.Sub(time.Now()) - threshold
+	if timeout < 0 || randTimeout < timeout {
+		timeout = randTimeout
+	}
+	timeoutSeconds := int64(timeout.Seconds())
+	return &timeoutSeconds
+}
+
+// Returns true if we are close enough to next planned periodic resync
+// and we can force resyncing ourself now.
+func (r *Reflector) canForceResyncNow() bool {
+	if r.nextResync.IsZero() {
+		return false
+	}
+	// Returns whether next resync is planned within 'threshold'.
+	threshold := 3 * time.Second
```

Comment:
Same thing about moving this to a variable and using a `now` func instead of `time.Now` directly.